toolchain/sem_ir/name_scope.h

```diff
@@ -18,13 +18,104 @@ enum class AccessKind : int8_t {
   Private,
 };
 
+// Represents the result of a name lookup.
+// A lookup result can be:
+// * Successful returning a valid `InstId`,
+// * Not found when the name wasn't declared or poisoned.
+// * Poisoned when the name wasn't declared but was poisoned.
+// * Represent that an error has occurred during lookup. This is still
+// considered found and the error `InstId` is considered valid.
+class ScopeLookupResult {
+ public:
+  // Creates a non poisoned result.
+  // Depending on the given `inst_id`, it can be:
+  // * Found: Lookup was successful and `inst_id` is valid. If `inst_id` is
+  // known to be valid, use `CreateFound()`.
+  // * Not Found: Lookup returned no results and `inst_id` is invalid. If
+  // `inst_id` is known to be invalid, use `CreateNotFound()`.
+  // * Error: Lookup encountered an error and `inst_id` is
+  // `ErrorInst::SingletonInstId`, which is considered valid and found. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  //
+  // To create a poisoned result, use `CreatePoisoned()`.
+  static auto Create(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    return ScopeLookupResult(inst_id, access_kind, false);
+  }
+
+  // Creates a successful lookup result. `inst_id` must be valid. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  static auto CreateFound(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    CARBON_CHECK(inst_id.is_valid());
+    return Create(inst_id, access_kind);
+  }
+
+  // Creates a result when there are no result.
+  static auto CreateNotFound() -> ScopeLookupResult {
+    return Create(InstId::Invalid, AccessKind::Public);
+  }
+
+  // Creates a poisoned result.
+  static auto CreatePoisoned() -> ScopeLookupResult {
+    return ScopeLookupResult(InstId::Invalid, AccessKind::Public, true);
+  }
+
+  // Creates a result when an error was encountered, which is considered valid
+  // and found.
+  static auto CreateError() -> ScopeLookupResult {
+    return CreateFound(ErrorInst::SingletonInstId, AccessKind::Public);
+  }
+
+  // True iff CreatePoisoned() was used.
+  auto is_poisoned() const -> bool { return is_poisoned_; }
+
+  // True when lookup was successful or resulted with an error.
+  // False for poisoned, not found.
+  auto is_found() const -> bool {
+    return !is_poisoned() && lookup_inst_id().is_valid();
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must not be called for poisoned results.
+  // If valid, it means the lookup was successful or resulted with an error.
+  auto lookup_inst_id() const -> InstId {
+    CARBON_CHECK(!is_poisoned());
+    return inst_id_;
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must only be called when lookup was successful.
```

Comment:
Can you apply paragraph wrapping to these comments?

Reply:
Done.